Windows support

dougthor42 opened the pull request:
This adds Windows support via [`colorama`](https://pypi.org/project/colorama/).

Fixes #270.

Includes #338. This PR should be rebased once that one is squashed and merged.

##Notes:

Due to the way `colorama` works - injecting win32 calls - I'm not sure if there's a way to test that colors are *actually* displayed. Had I opted to have `colorama` strip the ANSI codes, I could have used that to test that it's working. However, that would have meant updating the regex in at least 31 tests.


## Example:

```python
def fn():
    import pdb; pdb.set_trace()
    a = 1
    b = 2
    c = 3
    return a+b+c

fn()
```


### Before (v0.10.2)

#### Windows CMD
![image](https://user-images.githubusercontent.com/5386897/69499079-67135600-0ea3-11ea-8316-a95dce02537a.png)

#### PowerShell
![image](https://user-images.githubusercontent.com/5386897/69499109-af327880-0ea3-11ea-91ea-d65321112677.png)


### After

#### Windows CMD
![image](https://user-images.githubusercontent.com/5386897/69483788-8f864c00-0de0-11ea-94ce-4df709e82c1c.png)

![image](https://user-images.githubusercontent.com/5386897/69499437-fd954680-0ea6-11ea-94cc-0feaf34de15f.png)


#### PowerShell
![image](https://user-images.githubusercontent.com/5386897/69499147-f0c32380-0ea3-11ea-937f-b45c6882a64b.png)

![image](https://user-images.githubusercontent.com/5386897/69499153-046e8a00-0ea4-11ea-9503-a033ff572f04.png)

![image](https://user-images.githubusercontent.com/5386897/69499429-dd658780-0ea6-11ea-8c27-47034ea90088.png)



## Questions:


### `t_windows.py`

In blueyed/pdbpp@win2, you had a `t_windows.py` file. This appears to just be messing around with colorama - is that something you want to keep?


### `_setup_streams()`

Also in blueyed/pdbpp@win2, there was a `Pdb._setup_streams()` method:

```
def _setup_streams(self, stdout):
    self.stdout = self.ensure_file_can_write_unicode(stdout)

    if os.name == 'nt':
        try:
            import colorama
            supports_color = True
        except ImportError:
            supports_color = False
    else:
        supports_color = True

    # Setup highlight/use_pygments for detected color support.
    if self.config.highlight is None or self.config.use_pygments is None:
        if self.config.highlight is None:
            self.config.highlight = supports_color

    if os.name == "nt" and supports_color:
        if self.config.highlight or self.config.use_pygments is not False:
            self.stdout = colorama.AnsiToWin32(
                self.stdout, convert=True, strip=True
            ).stream
```

Do you want me to keep this for any reason? It's not needed for Windows support but you may have a refactoring in mind that I'm not aware of.


### Blink and Reveal in expected regex?

I'm failing `test_truncated_source_with_pygments_and_highlight`. It appears there's some wrapping around the `return` statement in the expected regex:

```
\^\[\[38;5;28;01;44mreturn\^\[\[39;00;44m
```

Those [ANSI codes](https://en.wikipedia.org/wiki/ANSI_escape_code#SGR_parameters) are
```
38: set foreground
5: slow blink
28: reveal
01: bold
44: background blue

39: default foreground
00: reset all
44: background blue
```

I don't understand why "slow blink" and "reveal" codes are in there. Are they added from pygments or something?

#### Failed Test Details

```
# Expected Regex:
\^\[\[30m\^\[\[47m *[0-9]+ +->         \^\[\[38;5;28;01;44mreturn\^\[\[39;00;44m a                                                       \^\[\[0{1,2}m
# Actual line:
^[[30m^[[47m1607  ->         return a                                                       ^[[0m
```

dougthor42 committed:
Initialize colorama at the start of the Pdb class.

Do not strip ANSI codes in colorama. This makes it so we don't have to
update all the test regex. ANSI codes don't do anything on Windows, so
it's safe to leave them in.

dougthor42 committed:
Import colorama if on windows. Print a warning if not present.

blueyed commented:
Took the liberty to rebase it myself (skipping a lot) already (to see how it goes, and so that you have something to compare to when doing it yourself also maybe).  Old HEAD: ab83498.
You might want to try if you come to the same result though.. :)

dougthor42 committed:
Add test_truncated_source_with_highlight

dougthor42 commented:
> Blink and Reveal in expected regex? `\^\[\[38;5;28;01;44mreturn\^\[\[39;00;44m`

I was mistaken. Those ANSI codes are actually [just colors](https://en.wikipedia.org/wiki/ANSI_escape_code#SGR_parameters):

```
38:   set foreground. Next arguments are "5;n" for 8-bit or "2;r;g;b" for 32-bit color
 5:   N/A. Part of the code for 38. Sets to 8-bit color.
28:   8-bit green-ish color.
```

That said, I think I know why `test_truncated_source_with_pygments_and_highlight` is failing: in 0831f0e90d49c46bd7c939dc73b16e8f5b96abb4 I added some code to `setbgcolor` that strips all formatting for the highlighted line.

Instead of removing the color formatting, I could try simulating the reversal instead (kinda like what was attempted in tartley/colorama#38), but then pretty much all of the "highlight" tests would need to be rewritten. Definitely want to avoid that. Hmm...

dougthor42 committed:
Fix current line highlighting on Windows.

dougthor42 committed:
Fix an issue with Windows linewrapping by decreasing the terminal width.

dougthor42 commented:
Alright, I think this is ready for a final review. I'm not sure what's going on with the failing py27 build - any thoughts there?

Turns out I was going about things all wrong initially, what with stripping colors in the current line. All that needed to be done is adjust the default line highlight color.

blueyed commented on src/pdbpp.py:
```diff
@@ -134,6 +144,10 @@ class DefaultConfig(object):
+    # Windows/colorama doesn't support 'inversed', so instead of using the
+    # terminal default colors, we just use black-on-white
+    if sys.platform == "win32":
+        current_line_color = "30;47"  # black, white
```
Does `49;39` work / make sense?  (poor man's inversed)

dougthor42 replied:
Sadly no, that doesn't make sense. The codes are independent of the location within the escape sequence. These are all the same "bold yellow text on blue background":
![image](https://user-images.githubusercontent.com/5386897/73961331-38f9bb80-48c1-11ea-937e-37827f6a6313.png)

With `49;39`:
![image](https://user-images.githubusercontent.com/5386897/73960459-ccca8800-48bf-11ea-8e13-c6b608131d81.png)

With `30;47`:
![image](https://user-images.githubusercontent.com/5386897/73960512-e9ff5680-48bf-11ea-81b5-5986f2d7b10c.png)

**If** there was a way to detect the current terminal's default background and foreground colors, then I could manually invert them. However, I'm not aware of any way to do that. It *may* be possible on Linux using [`curses`](https://docs.python.org/3/library/curses.html#curses.use_default_colors) (see [this SO answer](https://stackoverflow.com/a/44015131/1354930)), but I can't find anything on how to get the default colors in Windows CMD/PowerShell.

blueyed replied:
I've thought since `ESC[39;49m` sets/uses the default fg/bg this could be reversed maybe (just a silly thought.. :D)
As for what curses does (the SO answer) I guess it just skips setting it then, but have not checked actually.

I guess it is fine to use black on white then - although that might only work for dark schemes really.

After all it should / could use colors from the pygments theme/formatter maybe?

blueyed commented on src/pdbpp.py:
```diff
@@ -1646,6 +1664,21 @@ def get_terminal_size():
+            #   | 4         b = 2                            |
+            # So I figure we just fake the terminal size a bit. /shrug.
+            if has_colorama_on_windows:
+                width -= 1
```
Looks like it is related to current line (handling) then?  https://github.com/pdbpp/pdbpp/blob/cb56e1b6126bb87597684d01af6f18c873be85ee/src/pdbpp.py#L1102-L1115
Can you try without `current_line_color`?

dougthor42 replied:
Looks like this was a holdover of some other, now-unused changes and it looks like it's no longer needed. I've simply reverted it.

blueyed commented:
> I'm not sure what's going on with the failing py27 build - any thoughts there?

Works for me locally.  Looks similar to https://github.com/dougthor42/pdbpp/blob/c326745881a1db80f0b3df7e3c8a3bc209155812/testing/test_pdb.py#L3674-L3677, i.e. different line endings.

blueyed commented on src/pdbpp.py:
```diff
@@ -42,6 +42,16 @@ def signature(obj):
+        has_colorama_on_windows = True
+    except ImportError:
+        # No windows support. Do we want to stop trying all coloring?
+        print("Please install 'colorama' for color support on Windows.")
```
This should be done on demand, not during import - at least the printing.

dougthor42 replied:
Hmm... I'm not sure I fully understand. Are you saying I should put the whole thing (`sys.platform` check and the import) in `__init__` (L341-ish)?

blueyed replied:
Even later - only on demand, when there's actual output.

`import pdb; p = pdb.Pdb()` should not trigger an import of colorama.

dougthor42 replied:
Looks like the absolute latest that I can import and init `colorama` is:

```diff
@@ -352,6 +341,9 @@ class Pdb(pdb.Pdb, ConfigurableClass, object):
             self._disable_pytest_capture_maybe()
         kwargs = self.config.default_pdb_kwargs.copy()
         kwargs.update(**kwds)
+        if sys.platform == "win32":
+            import colorama
+            colorama.init(strip=False)
         super(Pdb, self).__init__(*args, **kwargs)
         self.prompt = self.config.prompt
         self.display_list = {}  # frame --> (name --> last seen value)
```

If I try importing colorama after that `super()`, then things don't work. I've pushed commit 4bd8cc1ff7f26704661990d1f115144aa8b7a1cb - take a look at let me know what you think.

I know if doesn't satisfy your requirement, but I don't know if there's a way to do so. If there is, it's outside of my skillset.

dougthor42 commented:
>> failing py27 build
>
> Works for me locally.

Yeah, same here. That'll make it annoying to debug 👿. I'll see what I can do.

blueyed commented:
> > > failing py27 build
> > Works for me locally.
> Yeah, same here. That'll make it annoying to debug . I'll see what I can do.

Good to know.
Don't let this block you, I can look into it.

dougthor42 committed:
Move the colorama import to as late as possible.

dougthor42 committed:
Revert "Fix an issue with Windows linewrapping by decreasing the terminal width."

This reverts commit acd2296242520017443b5a61cd70e21d1206f1c3.

dougthor42 committed:
Remove the 'has_colorama_on_windows' var - not longer needed.

emvalbuena commented:
Hi! 

Is there anything I can do to help having this merged?

I am having similar issues as reported in #270 with the Git Bash terminal. 

![image](https://user-images.githubusercontent.com/14164486/89172370-d0934a80-d582-11ea-9c20-fa3a7dd4d546.png)

The references to jupyter-like stuff corresponds to comment lines, this kind of prompt came up after I set a `import pdb; pdb.set_trace()` line in my script.

Thank you,

Enrique

blueyed commented:
@KikeM 
You could try if this PR / branch works for you, and report back.
IIRC the main issue to solve here is where to call the colorama hooks etc.

blueyed commented:
JFI: we've moved to GitHub Actions by now - which is pretty much relevant for this PR, so I wanted to let you know.. :)

asdf8601 commented:
I've tried and it works as expected!!

kurtmckee committed:
Update COLOR regexes in test_pdb.py:

+ Update COLORRESET to support colorama.
+ Adjust the COLORCURLINE regex to support colorama.
+ Split out the current line's 'linenumber' color formatting regex so that
  it can be used in test_sticky_dunder_return_with_highlight

sbor23 commented:
Can confirm it works as intended!
@dougthor42 @blueyed any plans to merge/release this? :)

dougthor42 commented:
Looks like I need to rebase onto latest master and fixup some tests. I'll try and get that done this weekend.

dougthor42 commented:
@blueyed I rebased this onto the current master as of today (098f728). There was a minor tweak with the new test but otherwise no significant changes.

Anything else you'd like me to do before this gets merged?

* The py27 job was canceled for some reason, but otherwise green across the board for tests.
* We never did resolve the fg/bg issue but I figured that could wait for later.
